### PR TITLE
[llvm-readobj][ELF] Implement JSON output for --dynamic-table

### PR DESCRIPTION
**Pull request by feg208**
When printing JSON output with --dynamic-table I noticed that the output is invalid JSON. This patch overrides the printDynamicTable() function in the JSONELFDumper to return a list of dictionaries for the DynamicSection value.
Before the output was:
```
 {
    "FileSummary": {
      "File": "bin/llvm-readelf",
      "Format": "elf64-x86-64",
      "Arch": "x86_64",
      "AddressSize": "64bit",
      "LoadName": "<Not found>"
    }DynamicSection [ (35 entries)
  Tag                Type         Name/Value
  0x000000000000001D RUNPATH      Library runpath: [$ORIGIN/../lib:]
  0x0000000000000001 NEEDED       Shared library: [libm.so.6]
  0x0000000000000001 NEEDED       Shared library: [libz.so.1]
  0x0000000000000001 NEEDED       Shared library: [libzstd.so.1]
```
Now the output looks like:
```
 "DynamicSection": [
      {
        "Tag": 29,
        "Type": "RUNPATH",
        "Value": 6322,
        "Path": [
          "$ORIGIN/../lib"
        ]
      },
      {
        "Tag": 1,
        "Type": "NEEDED",
        "Value": 6109,
        "Library": "libm.so.6"
      },
```

### CODE REVIEW

**Comment by ilovepi** [comment]
Can you include a snippet in the commit message that shows how the output is changing? We've done that for patches like this in the past, and it helps make clear what the new behavior is.

**Comment by ilovepi** [comment]
I think this is basically LGTM from me, modulo the output diff. I'll defer to @jh7370 for final approval, though.

**Comment by feg208** [comment]
> Can you include a snippet in the commit message that shows how the output is changing? We've done that for patches like this in the past, and it helps make clear what the new behavior is.

Done. I added it here and in the commit message

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:13]: As `-d` is an alias of `--dynamic-table`, and we have another test case that shows the behaviour of the alias, I don't see a need to repeat that check here (we have it for both llvm-readelf and llvm-readobj cases above, because the two have different command-lines).

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:12]: Could you check what other JSON output testing does, please as to whether it uses `--match-full-lines` and `--strict-whitespace`? I suspect the majority don't, so nothing to change here, but some minor changes can be made to the output test patterns (see below).

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:144]: Nit: the other check patterns don't include a space after the colon, because they need to match with `--match-full-lines --strict-whitespace`. If you're not doing the same here, I'd add that space, to match the common style in most tests (and one on each other line too for consistency).

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:152]: Nit: something weird here with the indentation.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7379]: I think we need to give this output a bit more consideration, given that this is JSON and therefore intended to be machine-readable (otherwise, we could just use LLVM-style output). In particular, I have the following two thoughts:
1) Tag/Type fields are just different ways of representing the same information. Do we need both of them?
2) The Value field is a string, yet this is almost never the "natural" representation of the value for a dynamic tag. It also is rarely the actual value, because it is an interpretation of that value. In my opinion, this potentially deserves splitting into two fields: a raw integer value (which is just the "bytes on disk" numeric value for the entry), and an interpreted value, in a format that is appropriate for the interpretation, e.g. a list of strings, or an integer (in the latter case, arguably the interpreted field may not be necessary).

Thoughts?

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: Done.

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: So I added --strict-whitespace back because once I fixed the indenting that passed fine.

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:151]: yeah. I cleaned this up

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: done

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:7379]: 1. I do like having the type string there but I can appreciate that it's duplicative.
2. Yeah. what about a construct like
```
"Value": { "bytes": 16 },
"Type": "PLTRELSZ",
},
....
"Value": { "Shared library": "[D]" }
"Type": "NEEDED"
},
...
"Value": {"value": 4096 },
"Type": "HASH"
},
```

and so on?

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:143]:
```suggestion
#      JSON64:   "DynamicSection": [
```
Do this and you still get the output nicely lined up for ease of reading.

That being said, I'd expect no spaces before "DynamicSection". IIRC, `--strict-whitespace` doesn't check leading and trailing whitespace without `--match-full-lines`, so I'm not sure there's a huge amount of point to having the former without the latter in this context (note that this is what the other cases do too). I'm also not 100% sure if it matters or not really, but I don't think it hurts to check them.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:7379]: I'm not opposed to having the type field, if you think it's useful for readability of the output.

As I understand JSON, I'm not sure we need a nested object for the Value field, because the individual DynamicTable entries can have different types for that field. In other words, I think it could look something like this:

```
...
"Value": 16
"Type": "PLTRELSZ"
...
"Value": ["A", "B"]
"RawValue":4321
"Type": "NEEDED"
...
"Value": 4096
"Type": "HASH"
```
Note that I've added "RawValue" for the case where the value isn't just the raw number in the dynamic table. For some dynamic entries, the value is the actual meaning. In other cases, it's a pointer to some other information stored somewhere. The "bytes" bit of the size is unnecessary, since the size values are implicitly in bytes.

A possible variation on my above suggestion is the following, where every entry has a "Value" member, which is the raw value stored in the dynamic table, and in the cases where that value has some interpreted meaning, there's an additional entry. In this way, a generic parser doesn't need to understand the tag types to be able to record the raw values, since all the "Value" entries have the same type.
```
"Type": "PLTRELSZ",
"Value": 16
...
"Type": "NEEDED",
"Value": 1234,
"Libraries": ["A", "B", "C"]
...
"Type": "SONAME",
"Value": 100,
"Name": "my-library"
```
(For simplicity in both examples, I've not included separate Tag/Type fields, but these can be included as per the current behaviour)

**Review by jh7370 — COMMENTED** [review]
Per the LLVM PR guidelines, please avoid force pushes on your branch: they make it harder to review, because you can no longer see what's changed since the previous version. If rebasing is necessary, you can use a merge commit from the base branch. When you squash and merge upon submission, the merge commit (and any fixup commits) will disappear.

I also prefer it if you don't "Resolve Conversation" on threads I've started, because I use this to track comments that I'm happy have been satisfactorily resolved (you'd be surprised how often people click "Resolve Conversation" when they haven't made the requested change, or they've made it incorrectly. See https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178 for more context.

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: I removed the spaces before "DynamicSection" and lined up that first line

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: I went with option 2. How does that strike you?

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (MaskRay) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:302]: trailing space may look strange

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (MaskRay) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7382]: AUXILIARY/FILTER are rarely used, but they are not "Libraries"

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (MaskRay) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7404]: Perhaps `return` for the default clause.

Add a `this->W.printString(this->getDynamicString(Entry.getVal()));` after the switch table for these string tags.

**Comment by MaskRay** [comment]
> Fix broken JSON output

The subject can probably say Implement JSON output for . The JSON output supports a few dump options and -d is not implemented yet.

**Comment by MaskRay** [comment]
> Per the LLVM PR guidelines, please avoid force pushes on your branch: they make it harder to review, because you can no longer see what's changed since the previous version. If rebasing is necessary, you can use a merge commit from the base branch. When you squash and merge upon submission, the merge commit (and any fixup commits) will disappear.

You can use getcord/spr if you prefer rebases and force pushes. The tool will propagate changes from your local branch to the PR branch and the PR branch only performs merges. https://maskray.me/blog/2023-09-09-reflections-on-llvm-switch-to-github-pull-requests#my-workflow

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:302]: Yeah. It seems like a bug in the dynamic string method. I'll find it and squash it. It doesn't show up in the LLVM formatter since there aren't quotes to delineate it

**Comment by feg208** [comment]
> You can use getcord/spr if you prefer rebases and force pushes. The tool will propagate changes from your local branch to the PR branch and the PR branch only performs merges. https://maskray.me/blog/2023-09-09-reflections-on-llvm-switch-to-github-pull-requests#my-workflow

Oh thanks a bunch. I'll check it out

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: Once we are good with the format I'll clean up the commit message so the output matches @ilovepi to keep everything tidy

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: Once we are good with the format I'll clean up the commit message so the output matches @ilovepi to keep everything tidy

**Comment by feg208** [comment]
> The subject can probably say Implement JSON output for . The JSON output supports a few dump options and -d is not implemented yet.

Done. Hopefully I did not give offense with the prior phrasing. None was meant certainly.

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:153]: Spacing has gone wrong again?

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7417]: Nit: as they're directly related, I'd put the Type string line immediately after the Tag one.

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:302]: I would have thought that Flags should be a list anyway?

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:398]: Same comment here as above: I feel like this would be better as a list, not a concatenated string.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7380]: I'm not sure `LLVM_FALLTHROUGH` is needed where there's no body to the case statement. You should be able to just do:
```
case DT_AUXILIARY:
case DT_FILTER:
case DT_NEEDED:
  ListScope L(this->W, "Libraries");
```

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7385]: Libraries is a list, so you should have 1 list element per library. You may want to enhance the test case too to have multiple elements.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7389]: Same comment as above, for here and all the other list types.

Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp:7389]: so reading the elf docs I think only the Path is really multiply valued. All the others are to an offset in the string table with no documented separator like ':' or ';' or anything. I wonder why the original was a
```
[D]
```
to begin with. I must be missing something though. I am going off https://refspecs.linuxfoundation.org/elf/elf.pdf 2-12

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: yeah. done

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: I am not super sure how to make this work? I tried doing a printString for the label and then a printString() below but that caused crashes. Could you give me a short example for what you are looking for here?

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: blech. vim. fixed

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: done

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test]: yeah I agree. done

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: changed

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (feg208) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: I mentioned this below but looking at the spec I am not clear how this can possibly have multiple elements. But there must be something I am missing

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [reply llvm/tools/llvm-readobj/ELFDumper.cpp]: Sorry, forgot to respond to this: current format is basically fine now, thanks. I'll continue reviewing and make any more comments I find.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7403]: Nit: I'd put the break inside the brace. Having it here feels wrong syntactically, even if it's not invalid.

File: llvm/tools/llvm-readobj/ELFDumper.cpp
Comment (jh7370) [inline llvm/tools/llvm-readobj/ELFDumper.cpp:7399]: Nit
```suggestion
      auto [Front, Back] = Value.split(':');
```

File: llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/dynamic-tags.test:221]: Here and for RUNPATH, you should probably have multiple elements in the path string. That will impact the output of the other test cases in this file, so I'm okay if you want to defer it to the follow-up patch I mention in the out-of-line comment, if you prefer.

**Review by jh7370 — APPROVED** [review]
LGTM now, with a couple of nits. Just don't forget to update the commit message before merging.

Would you be okay to do a (hopefully) small follow-up change to tidy up the slightly weird LLVM output, where it is using lists for single-item elements, please?

**Comment by feg208** [comment]
> LGTM now, with a couple of nits. Just don't forget to update the commit message before merging.

Will do.

> 
> Would you be okay to do a (hopefully) small follow-up change to tidy up the slightly weird LLVM output, where it is using lists for single-item elements, please?

Sure thing. Point me in the direction you are thinking and I'll make it happen

**Comment by jh7370** [comment]
> Sure thing. I'll make it happen. I am assuming just in this place.

As you've noticed yourself, it looks like the only case where there is a list is the RPATH/RUNPATH tags, so I guess take a look at `ListScope` occurrences in the relevant bit of code.

> > LGTM now, with a couple of nits. Just don't forget to update the commit message before merging.
> 
> Will do.
> 
> > Would you be okay to do a (hopefully) small follow-up change to tidy up the slightly weird LLVM output, where it is using lists for single-item elements, please?
> 
> Sure thing. I'll make it happen. I am assuming just in this place.

I'm thinking https://github.com/llvm/llvm-project/blob/db9e9eabb7835bae4285a3f13c7cc7c985455e27/llvm/tools/llvm-readobj/ELFDumper.cpp#L2474 is the area that needs changing for the LLVM-style output.

**Comment by feg208** [comment]
> I'm thinking
> 
> https://github.com/llvm/llvm-project/blob/db9e9eabb7835bae4285a3f13c7cc7c985455e27/llvm/tools/llvm-readobj/ELFDumper.cpp#L2474
> 
> is the area that needs changing for the LLVM-style output.

Do we want to just jettison the brackets? Or were you thinking of something else?

**Comment by jh7370** [comment]
> > I'm thinking
> > https://github.com/llvm/llvm-project/blob/db9e9eabb7835bae4285a3f13c7cc7c985455e27/llvm/tools/llvm-readobj/ELFDumper.cpp#L2474
> > 
> > is the area that needs changing for the LLVM-style output.
> 
> Do we want to just jettison the brackets? Or were you thinking of something else?

Just jettisoning the brackets. I suspect whoever implemented this option got themselves confused and tried to format things like a list for individual entries (note that it's certainly possible to have multiple DT_NEEDED etc tags, so a "summarised" version might have a list of needed libraries, but in the case we're talking about, we're only dealing with each tag individually, and therefore only a single entry).

**Pull request merged**
